### PR TITLE
[ie/francetv] Add support for episode and season tagging

### PR DESCRIPTION
**Pull request by Fymyte**
### Description of your *pull request* and other information

This adds support for episode and season tagging for FranceTV extractor.
The FranceTV extractor didn't used the fields `pre_title`, which contains season and episode number.
If those fields are filled, this means that the media is part of a TV Show and the Title is the Show name, and subtitle
is the episode title.

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check all of the following options that apply:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Fix or improvement to an extractor (Make sure to add/update tests)
- [ ] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

</details><details><summary>Copilot Summary</summary>  

### Summary
📺🇫🇷🔢

Added season and episode extraction for francetv series. Modified `yt_dlp/extractor/francetv.py` to parse `pre_title` metadata and output the corresponding fields.

> _`francetv` videos_
> _season and episode parsed_
> _autumn of series_

### Walkthrough
* Import the `re` module to use regular expressions for parsing season and episode numbers from the metadata ([link](https://github.com/yt-dlp/yt-dlp/pull/8409/files?diff=unified&w=0#diff-98bf13c62ffce8c8b537462644191b374f9d44df4fa242a6da9e7ce49cac14caR10))
* Initialize variables to store the episode number, season number, episode title and series title of the video, if available ([link](https://github.com/yt-dlp/yt-dlp/pull/8409/files?diff=unified&w=0#diff-98bf13c62ffce8c8b537462644191b374f9d44df4fa242a6da9e7ce49cac14caR86-R89))
* Extract the season and episode numbers from the `pre_title` field of the metadata, if present, using regular expressions ([link](https://github.com/yt-dlp/yt-dlp/pull/8409/files?diff=unified&w=0#diff-98bf13c62ffce8c8b537462644191b374f9d44df4fa242a6da9e7ce49cac14caL115-R127))
* Assign the episode title and series title to the `episode` and `series` variables, respectively, if the video is an episode from a series ([link](https://github.com/yt-dlp/yt-dlp/pull/8409/files?diff=unified&w=0#diff-98bf13c62ffce8c8b537462644191b374f9d44df4fa242a6da9e7ce49cac14caR206-R210))
* Include the `episode`, `series`, `episode_number` and `season_number` fields in the output dictionary of the extractor ([link](https://github.com/yt-dlp/yt-dlp/pull/8409/files?diff=unified&w=0#diff-98bf13c62ffce8c8b537462644191b374f9d44df4fa242a6da9e7ce49cac14caR224-R227))

</details>

### CODE REVIEW

**Review by seproDev — CHANGES_REQUESTED** [review]
Can you please also add a test?

File: yt_dlp/extractor/francetv.py [hunk 1]
```diff
@@ -112,7 +117,14 @@ def _extract_video(self, video_id, catalogue=None):
...
                     title = meta.get('title')
-                # XXX: what is meta['pre_title']?
+                # meta['pre_title'] contains season and episode number for series in format "S<ID> E<ID>"
+                if episode_number is None or season_number is None:
+                    pre_title = meta.get('pre_title')
+                    if pre_title:
+                        matches = re.findall(r'\d+', pre_title)
+                        if len(matches) >= 2:
+                            season_number = int(matches[0])
+                            episode_number = int(matches[1])
```
[the remark this answers is not in the document]
Comment (Fymyte) [reply yt_dlp/extractor/francetv.py]: Way more concise thx

File: yt_dlp/extractor/francetv.py [hunk 2]
```diff
@@ -191,19 +213,19 @@ def _extract_video(self, video_id, catalogue=None):
...
             })
 
-        if subtitle:
-            title += ' - %s' % subtitle
-        title = title.strip()
-
         return {
             'id': video_id,
-            'title': title,
+            'title': (title + ' - %s' % subtitle if subtitle else title).strip(),
```
Comment (seproDev) [inline yt_dlp/extractor/francetv.py:218]:
```suggestion
            'title': join_nonempty(title, subtitle, delim=' - ').strip(),
```
Import `join_nonempty` from `..utils`

**Review by seproDev — CHANGES_REQUESTED** [review]
Btw. you can apply suggestions right from the web interface.
In the future, please do not force push. All commits will be squashed upon merge anyway.

File: yt_dlp/extractor/francetv.py [hunk 3]
```diff
@@ -112,7 +132,9 @@ def _extract_video(self, video_id, catalogue=None):
             if meta:
                 if title is None:
                     title = meta.get('title')
-                # XXX: what is meta['pre_title']?
+                # meta['pre_title'] contains season and episode number for series in format "S<ID> E<ID>"
+                season_number, episode_number = self._search_regex(
+                        r'S(\d+)\s*E(\d+)', meta.get('pre_title'), 'episode info', group=(1, 2), default=(None, None))
```
[the remark this answers is not in the document]
Comment (Fymyte) [reply yt_dlp/extractor/francetv.py]: oups

**Comment by Fymyte** [comment]
Thx.  I didn't know about the above functions 👍. I knew for the apply in web interface but I had some extra changes to do.
I also had to change the previous download test for FranceTVSite because it was failing with the current content.

**Review by bashonly — APPROVED** [review]

**Pull request merged**